include/sys/dsl_scan.h

```diff
@@ -76,6 +77,18 @@ typedef enum dsl_scan_flags {
 
 #define	DSL_SCAN_FLAGS_MASK (DSF_VISIT_DS_AGAIN)
 
+typedef struct dsl_rescrub_phys {
+	uint64_t rescrub_func; /* pool_scan_func_t */
+	uint64_t rescrub_state; /* dsl_scan_state_t */
+	zap_cursor_t rescrub_cursor; /* zap cursor */
```

Comment:
For the `dsl_rescrub_phys_t` to be saved and correctly restored from the MOS as an array of uint64_t's, it can't include opaque types such as a `zap_cursor_t`.  Additionally, all members will need to be `uint64_t`'s (see rescrub_paused_flags below).

This is my first time looking at this proposed rescrub feature, so please correct me if I'm overlooking something critical.  But have you considered leveraging the the existing `dsl_scan_phys_t` type and the existing `dsl_scan` infrastructure (with a little modification).  It looks like the `dsl_scan_phys_t` contains everything you need.  You could introduce a new `DSF_SCRUB_ERROR_LOG` flag in `dsl_scan_flags_t` which would be set for rescrubs, then borrow one of the existing (now available) fields to store the `rescrub_cursor` progress as a `uint64_t`.  

Then I _think_ you'd potentially be able to take this a little farther and could remove the need for the new `pool_scrub_cmd_t`'s.  Which would probably make it possible to leverage the more of the existing scan infrastructure.  That would be nice since then I think you could avoid duplicating the start/pause/stop logic.

I'd be interested to hear if you looked in to this already, and would be happy to discuss this further.

Reply:
1. zap_cursor_t is getting stored in serialized form (`uint64_t's`). 
2. I agree most of the field are same. But I can't reuse the same data members in (`dsl_scan_phys_t`) because 1). Showing status for each scrub and error scrub in `zpool status` 2). If user pauses normal scrub and start error scrub then normal scrub information would get lost hence I do not think I can use the same data members.